Handle Solution Event UpdateSolution_QueryDelayFirstUpdateAction to delay build

jainaashish opened the pull request:
After discussion with Project System team, we now started listening to solution event UpdateSolution_QueryDelayFirstUpdateAction to delay build until any NuGet operation is already in progress without blocking UI thread. If NuGet operation is in progress, and some one hit build, then we set delay time to 1 sec which makes solution build manager system to call it again after 1 sec and keeps repeating it until we set it to 0. And we set it to 0 when our current NuGet operation is completed and releases NuGet lock.

It helps solve couple of our recent VS hangs where some NuGet operation is in progress like install package or auto restore, and at the same time, user trigger build so it will hang the VS. Now we won't block UI thread with this change so it won't hang the VS rather delay build until existing operation is completed.

Fixes https://github.com/NuGet/Home/issues/4420

@rrelyea @AArnott @emgarten @alpaix

Commit:
Started listening to UpdateSolution_QueryDelayFirstUpdateAction to delay build when NuGet operation is in progress

emgarten commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -72,6 +78,13 @@ public sealed class RestoreManagerPackage : AsyncPackage
                 _buildEvents.OnBuildBegin += BuildEvents_OnBuildBegin;
 
+                var solutionBuildManager = (IVsSolutionBuildManager5)await GetServiceAsync(typeof(SVsSolutionBuildManager));
+                if (solutionBuildManager != null)
```
Is it possible for this to be null? It might be better to fail if we can't hook up this event. We wouldn't know that future hangs aren't due to this getting missed.

AArnott replied:
In general, yes assume it could be null. New SKUs, combinations of workloads, etc., or a package load failure for that service could all cause it to be null. You can use `Assumes.Present(solutionBuildManager)` for this, and it will throw an exception with a reasonably helpful error message if it is null.

emgarten commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreWorker.cs:
```diff
@@ -386,6 +394,12 @@ public void CleanCache()
 
+            if (request.RestoreSource == RestoreOperationSource.Implicit && !IsAutoRestoreAllowed)
+            {
+                // don't run auto restore since onbuild restore is going to happen
```
Will auto restore cover all projects?

alpaix replied:
Don't think bailing the job is a good idea. Let's discuss this offline.

emgarten reviewed (changes requested):
Minor questions, overall this looks awesome :smile:

AArnott commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -72,6 +78,13 @@ public sealed class RestoreManagerPackage : AsyncPackage
+                var solutionBuildManager = (IVsSolutionBuildManager5)await GetServiceAsync(typeof(SVsSolutionBuildManager));
+                if (solutionBuildManager != null)
+                {
+                    uint updateSolutionEventsCookie4;
```
Please do not throw away the cookie. Store it in a field and unadvise in your package close method.

jainaashish replied:
This package will be there in whole VS session, and only be unloaded when you close VS so package Dispose method is never called which is why i didn't store this cookie and called into unadvise... there is no place i can call into unadvise method here.

AArnott replied:
Packages absolutely get disposed of on VS exit. Why do you think that doesn't happen?
There are occasionally memory leak runs that people check on VS exit, and advise calls without matching unadvise calls cause leaks to be reported. So even if you have it for the life of VS, it's good to balance your Advise calls with Unadvise calls.

AArnott commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -72,6 +78,13 @@ public sealed class RestoreManagerPackage : AsyncPackage
+                if (solutionBuildManager != null)
+                {
+                    uint updateSolutionEventsCookie4;
+                    solutionBuildManager.AdviseUpdateSolutionEvents4(this, out updateSolutionEventsCookie4);
```
Rather than drop the HRESULT, pass it to something like `ErrorHandler.ThrowOnFailure`

jainaashish replied:
`AdviseUpdateSolutionEvents4 `doesn't return any `HRESULT`

AArnott replied:
Oh.

AArnott commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -72,6 +78,13 @@ public sealed class RestoreManagerPackage : AsyncPackage
+                if (solutionBuildManager != null)
+                {
+                    uint updateSolutionEventsCookie4;
+                    solutionBuildManager.AdviseUpdateSolutionEvents4(this, out updateSolutionEventsCookie4);
```
Rather than drop the HRESULT, pass it to something like `ErrorHandler.ThrowOnFailure`

AArnott commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -114,5 +127,45 @@ private bool ShouldRestoreOnBuild
+        public void UpdateSolution_QueryDelayFirstUpdateAction(out int pfDelay)
+        {
+            // check if NuGet lock is already acquired by some other NuGet operation
+            if (LockService.IsLockHeld)
```
Does `IsLockHeld` return true when *someone else* is holding the lock, or just you?

jainaashish replied:
if someone is holding the lock...

alpaix replied:
That's redundant with the new `TryAcquireLock` approach

jainaashish replied:
yeah will remove it now

AArnott commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -114,5 +127,45 @@ private bool ShouldRestoreOnBuild
+            if (LockService.IsLockHeld)
+            {
+                // delay build by 1000ms and let it try again after that.
+                pfDelay = 1000;
```
Please change this to just assign `1`. You don't get to dictate the interval. It's just zero or non-zero. And so while setting `1000` is the same as `1` functionally, it gives the impression to the code reader that you get to set the length, and that is not the case.
I would also avoid specifying in your code comment that the delay is 1 second long since that is an implementation detail that can change.

jainaashish replied:
i see, makes sense!

AArnott commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -114,5 +127,45 @@ private bool ShouldRestoreOnBuild
+            else
+            {
+                // Disable running auto restore at this time, since it will start the on build restore
+                SolutionRestoreWorker.IsAutoRestoreAllowed = false;
```
Awesome.

AArnott commented on src/NuGet.Clients/PackageManagement.VisualStudio/SolutionRestore/ISolutionRestoreWorker.cs:
```diff
@@ -24,6 +24,11 @@ public interface ISolutionRestoreWorker
         bool IsBusy { get; }
 
         /// <summary>
+        /// Returnn true when auto restore can proceed without interrupting with onbuild restore
```
typo: `Returnn`

alpaix commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -27,7 +28,7 @@ namespace NuGet.SolutionRestoreManager
     [ProvideAutoLoad(VSConstants.UICONTEXT.SolutionExists_string)]
     [Guid(PackageGuidString)]
-    public sealed class RestoreManagerPackage : AsyncPackage
+    public sealed class RestoreManagerPackage : AsyncPackage, IVsUpdateSolutionEvents4
```
Why is this in the package? I'd suggest moving all the logic onto the worker. No need to split the responsibilities and define new contract.

jainaashish replied:
Discussed offline, it's required to be here since we want to Advise to these events as soon as solution loads.

AArnott replied:
I don't think that means it has to be here. In fact since other folks get pointers to your Package, this means they can cast that pointer to that interface and invoke you. That's probably not going to become a problem, but information hiding usually provides benefits where you don't expect them.
Just have a private nested class within this Package class (or another independent internal class) that implements this interface, new it up in your Package constructor and pass that instance to the Advise method.

alpaix reviewed (changes requested):
Let's re-evaluate the strategy offline.

jainaashish commented:
@alpaix @emgarten Based on our discussion, removed IsAllowedAutoRestore flag, instead made changes to acquire lock in this new solution event itself and release this lock once build is completed. Also added TryAcquireLock api so that build is never blocked to get NuGetLock, rather delayed and in return never block UI thread.

alpaix commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -114,5 +137,50 @@ private bool ShouldRestoreOnBuild
+            }
+            else
+            {
+                var isLockAcquired = LockService.TryAcquireLock(NuGetLockTimeout, out _nuGetLock);
```
As a pre-condition I'd suggest checking if `_nuGetLock` is null.

jainaashish replied:
why should we check it for null?

alpaix replied:
what if this is called multiple times and you already acquired the lock. You don't want to start acquiring another one.

jainaashish replied:
which is why I rather stick with `LockService.IsLockHeld` check instead of making `_nugetLock `null in dispose everytime and then checking for not null here... What you think?

alpaix replied:
I'm convinced the value should be checked for null even with apparently redundant first `if`. Let's hear some thoughts from other reviewers.

alpaix commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -114,5 +137,50 @@ private bool ShouldRestoreOnBuild
+        public void UpdateSolution_EndLastUpdateAction()
+        {
+            // Release NuGet lock so that other NuGet operation can continue now.
+            _nuGetLock?.Dispose();
```
Need to nullify the `_nuGetLock` after it's been disposed.

jainaashish replied:
Again how would that make a difference? Anyways it can't be used again after disposed besides it will always come here when it has actually acquired the lock and set `_nugetLock `to new IDisposible instance.

alpaix replied:
You assume these events are to be generated in nicely ordered reliable fashion. I suggest to add defensive conditions to guarantee lock is acquired and released only once. Does this make sense?

AArnott replied:
I won't get into the debate, except to say that I agree you should probably be as defensive as possible with regard to when you expect these events to be raised. There are so many kinds of builds (batch build, canceled builds, skipped builds, builds that are denied the right to even start, etc.) that each raise events in different ways, skipping event handlers, etc... And since you have basically no time to give this any road testing for escrow, I think you should be as defensive as possible.

alpaix commented on src/NuGet.Clients/PackageManagement.UI/Services/INuGetLockService.cs:
```diff
@@ -28,5 +28,13 @@ public interface INuGetLockService
+        /// <summary>
+        /// Try to obtain a lock with given timeout
+        /// </summary>
+        /// <param name="timeout">Waiting timoeout in millisecond.</param>
```
typo: _timeout_

rrelyea commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -72,6 +89,12 @@ public sealed class RestoreManagerPackage : AsyncPackage
                 _buildEvents = dte.Events.BuildEvents;
                 _buildEvents.OnBuildBegin += BuildEvents_OnBuildBegin;
 
+                var solutionBuildManager = (IVsSolutionBuildManager5)await GetServiceAsync(typeof(SVsSolutionBuildManager));
```
>    [](start = 5, length = 3)

consider putting the useragentstring before this block that you added, to be extra conservative.

AArnott commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -27,27 +29,39 @@ namespace NuGet.SolutionRestoreManager
+        /// Wait timeout for to acquire NuGet lock during build event. Don't block build if we can't 
+        /// acquire lock.
+        /// </summary>
+        private const int NuGetLockTimeout = 0;
```
I appreciate the xml doc comment on this member, but it doesn't describe the units (is this ms?) nor how `0` is interpreted (wait indefinitely, or wait for no time at all?). Consider using `private readonly static TimeSpan` for its type instead to remove all question as to its units.

AArnott commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs:
```diff
@@ -72,6 +89,12 @@ public sealed class RestoreManagerPackage : AsyncPackage
+                Assumes.Present(solutionBuildManager);
+
+                uint updateSolutionEventsCookie4;
+                solutionBuildManager.AdviseUpdateSolutionEvents4(this, out updateSolutionEventsCookie4);
```
I'd still prefer to see this output the cookie to a field rather than a local so you can unadvise in your package Close/Dispose method.

jainaashish commented:
Merged 109da1544ec8bface14a8982ce6d84f856049f9d
